xivo/pubsub.py

```diff
@@ -50,3 +53,37 @@ def unsubscribe(self, topic, callback):
 
         if not self._subscribers[topic]:
             self._subscribers.pop(topic, None)
+
+
+class CallbackCollector(object):
+
+    _TOPIC = 'callback-collector'
+
+    def __init__(self):
+        self._pubsub = Pubsub()
+        self._sources = set()
+        self._lock = threading.Lock()
+
+    def new_source(self):
+        source_id = uuid.uuid4()
+        logger.debug('Creating new callback source "%s"', source_id)
+        self._sources.add(source_id)
+        return partial(self.collect, source_id)
+
+    def collect(self, source_id, *args, **kwargs):
+        logger.debug('Collecting callback source "%s"', source_id)
+        if not self._sources:
+            logger.debug('Aborting collect')
+            return
+
+        with self._lock:
```

Comment:
Lock should be acquired before the if check, to avoid the following situation:
Collector: sources A and B
Thread A: collect (A)
Thread A: end collect(A)
Thread B: collect (B)
Thread C: collect (C) (assume there's a bug, and some thread tries to collect something that is not a source)
Thread B: check sources is not empty
Thread C: check sources is not empty
Thread B: acquire lock
Thread C: wait for lock
Thread B: publish() and release lock
Thread C: publish() and release lock (should do nothing)